File: LeetCode/543 - Diameter of Binary Tree.cpp
// Method1: diameter = max{every node's (lh + rh) while the path pass the node}
// Method2: run BFS twice

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    int diameter = 0;

    int height(TreeNode *root)
    {
        if (!root)
            return 0;

        int lh = height(root->left);
        int rh = height(root->right);

        // the path pass the node
        diameter = max(diameter, lh + rh);

        return 1 + max(lh, rh);
    }

    int diameterOfBinaryTree(TreeNode *root)
    {
        if (!root)
            return 0;

        height(root);
        return diameter;
    }
};

File: LeetCode/grind75/543 - Diameter of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = [0]
        self.dfs(root, res)
        return res[0]

    def dfs(self, root, res):
        if not root:
            return 0

        left = self.dfs(root.left, res)
        right = self.dfs(root.right, res)

        sum_depth = left + right
        if sum_depth > res[0]:
            res[0] = sum_depth

        if left > right:
            max_depth = left
        else:
            max_depth = right

        return max_depth + 1
